lighthouse-core/lib/sd-validation/schema-validator.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const walkObject = require('./helpers/walk-object.js');
+const schemaStructure = require('./assets/schema-tree.json');
+const TYPE_KEYWORD = '@type';
+const SCHEMA_ORG_URL_REGEX = /https?:\/\/schema\.org\//;
+
+/**
+ * @param {string} uri
+ * @returns {string}
+ */
+function cleanName(uri) {
+  return uri.replace(SCHEMA_ORG_URL_REGEX, '');
+}
+
+/**
+ * @param {string} type
+ * @returns {Array<string>}
+ */
+function getPropsForType(type) {
+  const cleanType = cleanName(type);
+  const props = schemaStructure.properties
+    .filter(prop => prop.parent.includes(cleanType))
+    .map(prop => prop.name);
+  const foundType = findType(type);
+  if (!foundType) throw new Error(`Unable to get props for missing type "${type}"`);
+  const parentTypes = foundType.parent;
+
+  return parentTypes.reduce((allProps, type) => allProps.concat(getPropsForType(type)), props);
+}
+
+/**
+ * @param {string} type
+ * @returns {{name: string, parent: Array<string>}|undefined}
+ */
+function findType(type) {
+  const cleanType = cleanName(type);
+
+  return schemaStructure.types.find(typeObj => typeObj.name === cleanType);
+}
+
+/**
+ * Validates keys of given object based on its type(s). Returns an array of error messages.
+ *
+ * @param {string|Array<string>} typeOrTypes
+ * @param {Array<string>} keys
+ * @returns {Array<string>}
+ */
+function validateObjectKeys(typeOrTypes, keys) {
+  /** @type {Array<string>} */
+  let types = [];
+
+  if (typeof typeOrTypes === 'string') {
+    types = [typeOrTypes];
+  } else if (Array.isArray(typeOrTypes)) {
+    types = typeOrTypes;
```

Comment:
Should we be ensuring that all the elements are also 'string'?

Reply:
"If the node object contains the @type key, its value must be either an absolute IRI, a relative IRI, a compact IRI (including blank node identifiers), a term defined in the active context expanding into an absolute IRI, or an array of any of these. See section 5.4 Specifying the Type for further discussion on @type values."

Reply:
Sure sounds good 👍